// @capacitor-firebase/firestore/dist/plugin.cjs.js
'use strict';

var core = require('@capacitor/core');
var app = require('firebase/app');
var firestore = require('firebase/firestore');

class DocumentReference {
    constructor(path) {
        DocumentReference.validatePath(path);
        this.path = path;
        this.id = path.substring(path.lastIndexOf('/') + 1);
    }
    static fromPath(path) {
        return new DocumentReference(path);
    }
    toJSON() {
        return {
            __type__: 'documentReference',
            id: this.id,
            path: this.path,
        };
    }
    static validatePath(path) {
        const segments = path.split('/');
        const hasEmptySegment = segments.some(segment => segment.length === 0);
        if (segments.length % 2 !== 0 || hasEmptySegment) {
            throw new Error(`Invalid document reference. Document references must point to a document with an even number of non-empty path segments, but got '${path}'.`);
        }
    }
}

class Bytes {
    constructor(base64) {
        this.base64 = base64;
    }
    static fromBase64String(base64) {
        return new Bytes(base64);
    }
    static fromUint8Array(array) {
        const chunkSize = 0x8000;
        const chunks = [];
        for (let i = 0; i < array.byteLength; i += chunkSize) {
            const chunk = array.subarray(i, i + chunkSize);
            chunks.push(String.fromCharCode.apply(null, chunk));
        }
        return new Bytes(btoa(chunks.join('')));
    }
    toBase64() {
        return this.base64;
    }
    toUint8Array() {
        const binary = atob(this.base64);
        const array = new Uint8Array(binary.length);
        for (let i = 0; i < binary.length; i++) {
            array[i] = binary.charCodeAt(i);
        }
        return array;
    }
    toJSON() {
        return {
            __type__: 'bytes',
            bytes: this.base64,
        };
    }
}

class FieldValue {
    constructor(marker) {
        this.marker = marker;
    }
    static serverTimestamp() {
        return new FieldValue({ __type__: 'serverTimestamp' });
    }
    static arrayUnion(...elements) {
        return new FieldValue({ __type__: 'arrayUnion', elements });
    }
    static arrayRemove(...elements) {
        return new FieldValue({ __type__: 'arrayRemove', elements });
    }
    static delete() {
        return new FieldValue({ __type__: 'delete' });
    }
    static increment(operand) {
        return new FieldValue({ __type__: 'increment', operand });
    }
    toJSON() {
        return Object.assign({}, this.marker);
    }
}

class GeoPoint {
    constructor(latitude, longitude) {
        this.latitude = latitude;
        this.longitude = longitude;
    }
    toJSON() {
        return {
            __type__: 'geopoint',
            latitude: this.latitude,
            longitude: this.longitude,
        };
    }
}

function serializeSpecialNumber(value) {
    if (Number.isNaN(value)) {
        return { __type__: 'number', value: 'NaN' };
    }
    if (value === Infinity) {
        return { __type__: 'number', value: 'Infinity' };
    }
    return { __type__: 'number', value: '-Infinity' };
}
function deserializeSpecialNumber(value) {
    if (value === 'NaN') {
        return NaN;
    }
    if (value === 'Infinity') {
        return Infinity;
    }
    if (value === '-Infinity') {
        return -Infinity;
    }
    return Number(value);
}

class Timestamp {
    constructor(seconds, nanoseconds) {
        this.seconds = seconds;
        this.nanoseconds = nanoseconds;
    }
    static now() {
        return Timestamp.fromMillis(Date.now());
    }
    static fromDate(date) {
        const ms = date.getTime();
        const seconds = Math.floor(ms / 1000);
        const nanoseconds = (ms % 1000) * 1000000;
        return new Timestamp(seconds, nanoseconds);
    }
    static fromMillis(ms) {
        const seconds = Math.floor(ms / 1000);
        const nanoseconds = (ms % 1000) * 1000000;
        return new Timestamp(seconds, nanoseconds);
    }
    toDate() {
        return new Date(this.toMillis());
    }
    toMillis() {
        return this.seconds * 1000 + Math.floor(this.nanoseconds / 1000000);
    }
    toJSON() {
        return {
            __type__: 'timestamp',
            seconds: this.seconds,
            nanoseconds: this.nanoseconds,
        };
    }
}

function serializeData(data) {
    if (data === null || data === undefined) {
        return data;
    }
    if (typeof data === 'number' && !Number.isFinite(data)) {
        return serializeSpecialNumber(data);
    }
    if (data instanceof Timestamp ||
        data instanceof GeoPoint ||
        data instanceof Bytes ||
        data instanceof DocumentReference ||
        data instanceof FieldValue) {
        return data.toJSON();
    }
    if (Array.isArray(data)) {
        return data.map(item => serializeData(item));
    }
    if (typeof data === 'object') {
        const result = {};
        for (const key of Object.keys(data)) {
            result[key] = serializeData(data[key]);
        }
        return result;
    }
    return data;
}
function deserializeData(data) {
    if (data === null || data === undefined) {
        return data;
    }
    if (Array.isArray(data)) {
        return data.map(item => deserializeData(item));
    }
    if (typeof data === 'object') {
        if (data.__type__ === 'timestamp') {
            return new Timestamp(data.seconds, data.nanoseconds);
        }
        if (data.__type__ === 'geopoint') {
            return new GeoPoint(data.latitude, data.longitude);
        }
        if (data.__type__ === 'bytes') {
            return Bytes.fromBase64String(data.bytes);
        }
        if (data.__type__ === 'documentReference') {
            return DocumentReference.fromPath(data.path);
        }
        if (data.__type__ === 'number') {
            return deserializeSpecialNumber(data.value);
        }
        const result = {};
        for (const key of Object.keys(data)) {
            result[key] = deserializeData(data[key]);
        }
        return result;
    }
    return data;
}

class FirebaseFirestoreClient {
    constructor(plugin) {
        this.plugin = plugin;
    }
    async addCollectionGroupSnapshotListener(options, callback) {
        return this.plugin.addCollectionGroupSnapshotListener(serializeData(options), (event, error) => {
            if (event) {
                for (const snapshot of event.snapshots) {
                    if (snapshot.data !== null) {
                        snapshot.data = deserializeData(snapshot.data);
                    }
                }
            }
            callback(event, error);
        });
    }
    async addCollectionSnapshotListener(options, callback) {
        return this.plugin.addCollectionSnapshotListener(serializeData(options), (event, error) => {
            if (event) {
                for (const snapshot of event.snapshots) {
                    if (snapshot.data !== null) {
                        snapshot.data = deserializeData(snapshot.data);
                    }
                }
            }
            callback(event, error);
        });
    }
    async addDocument(options) {
        const result = await this.plugin.addDocument(Object.assign(Object.assign({}, options), { data: serializeData(options.data) }));
        result.reference = DocumentReference.fromPath(result.reference.path);
        return result;
    }
    async addDocumentSnapshotListener(options, callback) {
        return this.plugin.addDocumentSnapshotListener(options, (event, error) => {
            if ((event === null || event === void 0 ? void 0 : event.snapshot.data) !== null &&
                (event === null || event === void 0 ? void 0 : event.snapshot.data) !== undefined) {
                event.snapshot.data = deserializeData(event.snapshot.data);
            }
            callback(event, error);
        });
    }
    async clearPersistence() {
        return this.plugin.clearPersistence();
    }
    async deleteDocument(options) {
        return this.plugin.deleteDocument(options);
    }
    async disableNetwork() {
        return this.plugin.disableNetwork();
    }
    async disablePersistence() {
        return this.plugin.disablePersistence();
    }
    async enableNetwork() {
        return this.plugin.enableNetwork();
    }
    async enablePersistence(options) {
        return this.plugin.enablePersistence(options);
    }
    async getCollection(options) {
        const result = await this.plugin.getCollection(serializeData(options));
        for (const snapshot of result.snapshots) {
            if (snapshot.data !== null) {
                snapshot.data = deserializeData(snapshot.data);
            }
        }
        return result;
    }
    async getCollectionGroup(options) {
        const result = await this.plugin.getCollectionGroup(serializeData(options));
        for (const snapshot of result.snapshots) {
            if (snapshot.data !== null) {
                snapshot.data = deserializeData(snapshot.data);
            }
        }
        return result;
    }
    async getCountFromServer(options) {
        return this.plugin.getCountFromServer(options);
    }
    async getDocument(options) {
        const result = await this.plugin.getDocument(options);
        if (result.snapshot.data !== null) {
            result.snapshot.data = deserializeData(result.snapshot.data);
        }
        return result;
    }
    async removeAllListeners() {
        return this.plugin.removeAllListeners();
    }
    async removeSnapshotListener(options) {
        return this.plugin.removeSnapshotListener(options);
    }
    async setDocument(options) {
        return this.plugin.setDocument(Object.assign(Object.assign({}, options), { data: serializeData(options.data) }));
    }
    async updateDocument(options) {
        return this.plugin.updateDocument(Object.assign(Object.assign({}, options), { data: serializeData(options.data) }));
    }
    async useEmulator(options) {
        return this.plugin.useEmulator(options);
    }
    async writeBatch(options) {
        return this.plugin.writeBatch(Object.assign(Object.assign({}, options), { operations: options.operations.map(op => (Object.assign(Object.assign({}, op), { data: op.data ? serializeData(op.data) : op.data }))) }));
    }
}

const plugin = core.registerPlugin('FirebaseFirestore', {
    web: () => Promise.resolve().then(function () { return web; }).then(m => new m.FirebaseFirestoreWeb()),
});
const FirebaseFirestore = new FirebaseFirestoreClient(plugin);

function normalizeServerTimestamps(value) {
    return value === 'estimate' || value === 'previous' ? value : 'none';
}
class FirebaseFirestoreWeb extends core.WebPlugin {
    constructor() {
        super(...arguments);
        this.unsubscribesMap = new Map();
        this.lastListenerId = 0;
    }
    async addCollectionGroupSnapshotListener(options, callback) {
        const collectionQuery = await this.buildCollectionQuery(options, 'collectionGroup');
        const unsubscribe = firestore.onSnapshot(collectionQuery, {
            includeMetadataChanges: options.includeMetadataChanges,
            source: options.source,
        }, snapshot => {
            const event = {
                snapshots: snapshot.docs.map(documentSnapshot => ({
                    id: documentSnapshot.id,
                    path: documentSnapshot.ref.path,
                    data: this.deserializeData(documentSnapshot.data({
                        serverTimestamps: normalizeServerTimestamps(options.serverTimestamps),
                    })),
                    metadata: {
                        hasPendingWrites: documentSnapshot.metadata.hasPendingWrites,
                        fromCache: documentSnapshot.metadata.fromCache,
                    },
                })),
            };
            callback(event, undefined);
        }, error => callback(null, error));
        const id = this.generateListenerId();
        this.unsubscribesMap.set(id, unsubscribe);
        return id;
    }
    async addCollectionSnapshotListener(options, callback) {
        const collectionQuery = await this.buildCollectionQuery(options, 'collection');
        const unsubscribe = firestore.onSnapshot(collectionQuery, {
            includeMetadataChanges: options.includeMetadataChanges,
            source: options.source,
        }, snapshot => {
            const event = {
                snapshots: snapshot.docs.map(documentSnapshot => ({
                    id: documentSnapshot.id,
                    path: documentSnapshot.ref.path,
                    data: this.deserializeData(documentSnapshot.data({
                        serverTimestamps: normalizeServerTimestamps(options.serverTimestamps),
                    })),
                    metadata: {
                        hasPendingWrites: documentSnapshot.metadata.hasPendingWrites,
                        fromCache: documentSnapshot.metadata.fromCache,
                    },
                })),
            };
            callback(event, undefined);
        }, error => callback(null, error));
        const id = this.generateListenerId();
        this.unsubscribesMap.set(id, unsubscribe);
        return id;
    }
    async addDocument(options) {
        const firestore$1 = firestore.getFirestore();
        const { reference, data } = options;
        const documentReference = await firestore.addDoc(firestore.collection(firestore$1, reference), this.serializeData(data));
        return {
            reference: DocumentReference.fromPath(documentReference.path),
        };
    }
    async addDocumentSnapshotListener(options, callback) {
        const firestore$1 = firestore.getFirestore();
        const unsubscribe = firestore.onSnapshot(firestore.doc(firestore$1, options.reference), {
            includeMetadataChanges: options.includeMetadataChanges,
            source: options.source,
        }, snapshot => {
            const data = snapshot.data({
                serverTimestamps: normalizeServerTimestamps(options.serverTimestamps),
            });
            const event = {
                snapshot: {
                    id: snapshot.id,
                    path: snapshot.ref.path,
                    data: (data === undefined
                        ? null
                        : this.deserializeData(data)),
                    metadata: {
                        hasPendingWrites: snapshot.metadata.hasPendingWrites,
                        fromCache: snapshot.metadata.fromCache,
                    },
                },
            };
            callback(event, undefined);
        }, error => callback(null, error));
        const id = this.generateListenerId();
        this.unsubscribesMap.set(id, unsubscribe);
        return id;
    }
    async clearPersistence() {
        const firestore$1 = firestore.getFirestore();
        await firestore.clearIndexedDbPersistence(firestore$1);
    }
    async deleteDocument(options) {
        const firestore$1 = firestore.getFirestore();
        const { reference } = options;
        await firestore.deleteDoc(firestore.doc(firestore$1, reference));
    }
    async disableNetwork() {
        const firestore$1 = firestore.getFirestore();
        await firestore.disableNetwork(firestore$1);
    }
    async disablePersistence() {
        firestore.initializeFirestore(app.getApp(), {
            localCache: firestore.memoryLocalCache(),
        });
    }
    async enableNetwork() {
        const firestore$1 = firestore.getFirestore();
        await firestore.enableNetwork(firestore$1);
    }
    async enablePersistence(options) {
        const tabManager = (options === null || options === void 0 ? void 0 : options.synchronizeTabs)
            ? firestore.persistentMultipleTabManager()
            : firestore.persistentSingleTabManager(undefined);
        firestore.initializeFirestore(app.getApp(), {
            localCache: firestore.persistentLocalCache({
                tabManager,
                cacheSizeBytes: options === null || options === void 0 ? void 0 : options.cacheSizeBytes,
            }),
        });
    }
    async getCollection(options) {
        const collectionQuery = await this.buildCollectionQuery(options, 'collection');
        const collectionSnapshot = await firestore.getDocs(collectionQuery);
        return {
            snapshots: collectionSnapshot.docs.map(documentSnapshot => ({
                id: documentSnapshot.id,
                path: documentSnapshot.ref.path,
                data: this.deserializeData(documentSnapshot.data()),
                metadata: {
                    hasPendingWrites: documentSnapshot.metadata.hasPendingWrites,
                    fromCache: documentSnapshot.metadata.fromCache,
                },
            })),
        };
    }
    async getCollectionGroup(options) {
        const collectionQuery = await this.buildCollectionQuery(options, 'collectionGroup');
        const collectionSnapshot = await firestore.getDocs(collectionQuery);
        return {
            snapshots: collectionSnapshot.docs.map(documentSnapshot => ({
                id: documentSnapshot.id,
                path: documentSnapshot.ref.path,
                data: this.deserializeData(documentSnapshot.data()),
                metadata: {
                    hasPendingWrites: documentSnapshot.metadata.hasPendingWrites,
                    fromCache: documentSnapshot.metadata.fromCache,
                },
            })),
        };
    }
    async getCountFromServer(options) {
        const collectionQuery = await this.buildCollectionQuery(options, 'collection');
        const snapshot = await firestore.getCountFromServer(collectionQuery);
        return { count: snapshot.data().count };
    }
    async getDocument(options) {
        const firestore$1 = firestore.getFirestore();
        const { reference } = options;
        const documentSnapshot = await firestore.getDoc(firestore.doc(firestore$1, reference));
        const documentSnapshotData = documentSnapshot.data();
        return {
            snapshot: {
                id: documentSnapshot.id,
                path: documentSnapshot.ref.path,
                data: (documentSnapshotData === undefined
                    ? null
                    : this.deserializeData(documentSnapshotData)),
                metadata: {
                    hasPendingWrites: documentSnapshot.metadata.hasPendingWrites,
                    fromCache: documentSnapshot.metadata.fromCache,
                },
            },
        };
    }
    async removeAllListeners() {
        this.unsubscribesMap.forEach(unsubscribe => unsubscribe());
        this.unsubscribesMap.clear();
        await super.removeAllListeners();
    }
    async removeSnapshotListener(options) {
        const unsubscribe = this.unsubscribesMap.get(options.callbackId);
        if (!unsubscribe) {
            return;
        }
        unsubscribe();
        this.unsubscribesMap.delete(options.callbackId);
    }
    async setDocument(options) {
        const firestore$1 = firestore.getFirestore();
        const { reference, data, merge } = options;
        await firestore.setDoc(firestore.doc(firestore$1, reference), this.serializeData(data), {
            merge,
        });
    }
    async updateDocument(options) {
        const firestore$1 = firestore.getFirestore();
        const { reference, data } = options;
        await firestore.updateDoc(firestore.doc(firestore$1, reference), this.serializeData(data));
    }
    async useEmulator(options) {
        const firestore$1 = firestore.getFirestore();
        const port = options.port || 8080;
        firestore.connectFirestoreEmulator(firestore$1, options.host, port);
    }
    async writeBatch(options) {
        const firestore$1 = firestore.getFirestore();
        const { operations } = options;
        const batch = firestore.writeBatch(firestore$1);
        for (const operation of operations) {
            const { type, reference, data, options } = operation;
            const documentReference = firestore.doc(firestore$1, reference);
            switch (type) {
                case 'set':
                    batch.set(documentReference, this.serializeData(data !== null && data !== void 0 ? data : {}), options !== null && options !== void 0 ? options : {});
                    break;
                case 'update':
                    batch.update(documentReference, this.serializeData(data !== null && data !== void 0 ? data : {}));
                    break;
                case 'delete':
                    batch.delete(documentReference);
                    break;
            }
        }
        await batch.commit();
    }
    async buildCollectionQuery(options, type) {
        const firestore$1 = firestore.getFirestore();
        let collectionQuery;
        if (options.compositeFilter) {
            const compositeFilter = this.buildFirebaseQueryCompositeFilterConstraint(options.compositeFilter);
            const queryConstraints = await this.buildFirebaseQueryNonFilterConstraints(options.queryConstraints || []);
            collectionQuery = firestore.query(type === 'collection'
                ? firestore.collection(firestore$1, options.reference)
                : firestore.collectionGroup(firestore$1, options.reference), compositeFilter, ...queryConstraints);
        }
        else {
            const queryConstraints = await this.buildFirebaseQueryConstraints(options.queryConstraints || []);
            collectionQuery = firestore.query(type === 'collection'
                ? firestore.collection(firestore$1, options.reference)
                : firestore.collectionGroup(firestore$1, options.reference), ...queryConstraints);
        }
        return collectionQuery;
    }
    buildFirebaseQueryCompositeFilterConstraint(compositeFilter) {
        const queryConstraints = this.buildFirebaseQueryFilterConstraints(compositeFilter.queryConstraints);
        if (compositeFilter.type === 'and') {
            return firestore.and(...queryConstraints);
        }
        else {
            return firestore.or(...queryConstraints);
        }
    }
    async buildFirebaseQueryConstraint(queryConstraint) {
        if (queryConstraint.type === 'where') {
            return this.buildFirebaseQueryFieldFilterConstraint(queryConstraint);
        }
        else {
            return await this.buildFirebaseQueryNonFilterConstraint(queryConstraint);
        }
    }
    async buildFirebaseQueryConstraints(queryConstraints) {
        const firebaseQueryConstraints = [];
        for (const queryConstraint of queryConstraints) {
            const firebaseQueryConstraint = await this.buildFirebaseQueryConstraint(queryConstraint);
            firebaseQueryConstraints.push(firebaseQueryConstraint);
        }
        return firebaseQueryConstraints;
    }
    buildFirebaseQueryFieldFilterConstraint(queryfilterConstraints) {
        return firestore.where(queryfilterConstraints.fieldPath, queryfilterConstraints.opStr, this.serializeFilterValue(queryfilterConstraints.value));
    }
    buildFirebaseQueryFilterConstraint(queryFilterConstraints) {
        if (queryFilterConstraints.type === 'where') {
            return this.buildFirebaseQueryFieldFilterConstraint(queryFilterConstraints);
        }
        else {
            return this.buildFirebaseQueryCompositeFilterConstraint(queryFilterConstraints);
        }
    }
    buildFirebaseQueryFilterConstraints(queryfilterConstraints) {
        const firebaseQueryFilterConstraints = [];
        for (const queryfilterConstraint of queryfilterConstraints) {
            const firebaseQueryFilterConstraint = this.buildFirebaseQueryFilterConstraint(queryfilterConstraint);
            firebaseQueryFilterConstraints.push(firebaseQueryFilterConstraint);
        }
        return firebaseQueryFilterConstraints;
    }
    async buildFirebaseQueryNonFilterConstraint(queryConstraints) {
        switch (queryConstraints.type) {
            case 'orderBy':
                return firestore.orderBy(queryConstraints.fieldPath, queryConstraints.directionStr);
            case 'limit':
                return firestore.limit(queryConstraints.limit);
            case 'limitToLast':
                return firestore.limitToLast(queryConstraints.limit);
            case 'startAt':
            case 'startAfter':
            case 'endAt':
            case 'endBefore': {
                const firestore$1 = firestore.getFirestore();
                const documentSnapshot = await firestore.getDoc(firestore.doc(firestore$1, queryConstraints.reference));
                switch (queryConstraints.type) {
                    case 'startAt':
                        return firestore.startAt(documentSnapshot);
                    case 'startAfter':
                        return firestore.startAfter(documentSnapshot);
                    case 'endAt':
                        return firestore.endAt(documentSnapshot);
                    case 'endBefore':
                        return firestore.endBefore(documentSnapshot);
                }
            }
        }
    }
    async buildFirebaseQueryNonFilterConstraints(queryConstraints) {
        const firebaseQueryNonFilterConstraints = [];
        for (const queryConstraint of queryConstraints) {
            const firebaseQueryNonFilterConstraint = await this.buildFirebaseQueryNonFilterConstraint(queryConstraint);
            firebaseQueryNonFilterConstraints.push(firebaseQueryNonFilterConstraint);
        }
        return firebaseQueryNonFilterConstraints;
    }
    deserializeData(data) {
        if (data === null || data === undefined) {
            return data;
        }
        if (data instanceof firestore.Timestamp) {
            return {
                __type__: 'timestamp',
                seconds: data.seconds,
                nanoseconds: data.nanoseconds,
            };
        }
        if (data instanceof firestore.GeoPoint) {
            return {
                __type__: 'geopoint',
                latitude: data.latitude,
                longitude: data.longitude,
            };
        }
        if (data instanceof firestore.DocumentReference) {
            return {
                __type__: 'documentReference',
                id: data.id,
                path: data.path,
            };
        }
        if (data instanceof firestore.Bytes) {
            return {
                __type__: 'bytes',
                bytes: data.toBase64(),
            };
        }
        if (typeof data === 'number' && !Number.isFinite(data)) {
            return serializeSpecialNumber(data);
        }
        if (Array.isArray(data)) {
            return data.map(item => this.deserializeData(item));
        }
        if (typeof data === 'object') {
            const result = {};
            for (const key of Object.keys(data)) {
                result[key] = this.deserializeData(data[key]);
            }
            return result;
        }
        return data;
    }
    serializeData(data) {
        if (data instanceof Timestamp) {
            return new firestore.Timestamp(data.seconds, data.nanoseconds);
        }
        if (data instanceof GeoPoint) {
            return new firestore.GeoPoint(data.latitude, data.longitude);
        }
        if (data instanceof Bytes) {
            return firestore.Bytes.fromBase64String(data.toBase64());
        }
        if (data instanceof FieldValue) {
            return this.serializeFieldValue(data.toJSON());
        }
        if (data === null || data === undefined) {
            return data;
        }
        if (Array.isArray(data)) {
            return data.map(item => this.serializeData(item));
        }
        if (typeof data === 'object') {
            if (data.__type__) {
                return this.serializeMarker(data);
            }
            const result = {};
            for (const key of Object.keys(data)) {
                result[key] = this.serializeData(data[key]);
            }
            return result;
        }
        return data;
    }
    serializeFieldValue(marker) {
        return this.serializeMarker(marker);
    }
    serializeFilterValue(value) {
        return this.serializeData(value);
    }
    serializeMarker(marker) {
        switch (marker.__type__) {
            case 'timestamp':
                return new firestore.Timestamp(marker.seconds, marker.nanoseconds);
            case 'geopoint':
                return new firestore.GeoPoint(marker.latitude, marker.longitude);
            case 'documentReference':
                return firestore.doc(firestore.getFirestore(), marker.path);
            case 'bytes':
                return firestore.Bytes.fromBase64String(marker.bytes);
            case 'number':
                return deserializeSpecialNumber(marker.value);
            case 'serverTimestamp':
                return firestore.serverTimestamp();
            case 'arrayUnion':
                return firestore.arrayUnion(...(marker.elements || []).map((e) => this.serializeData(e)));
            case 'arrayRemove':
                return firestore.arrayRemove(...(marker.elements || []).map((e) => this.serializeData(e)));
            case 'delete':
                return firestore.deleteField();
            case 'increment':
                return firestore.increment(marker.operand);
            default:
                return marker;
        }
    }
    generateListenerId() {
        return (++this.lastListenerId).toString();
    }
}

var web = /*#__PURE__*/Object.freeze({
    __proto__: null,
    FirebaseFirestoreWeb: FirebaseFirestoreWeb
});

exports.Bytes = Bytes;
exports.DocumentReference = DocumentReference;
exports.FieldValue = FieldValue;
exports.FirebaseFirestore = FirebaseFirestore;
exports.GeoPoint = GeoPoint;
exports.Timestamp = Timestamp;
